paddle/fluid/framework/scope.cc

```diff
@@ -200,6 +200,21 @@ Variable* Scope::FindVarLocally(const std::string& name) const {
   return nullptr;
 }
 
+void Scope::ClearWithPreserve(const std::unordered_set<Variable*>& vars) {
+  DropKids();
+
+  {
+    SCOPE_VARS_WRITER_LOCK
+    for (auto iter = vars_.begin(); iter != vars_.end();) {
+      if (vars.count(iter->second.get()) != 0) {
+        ++iter;
+      } else {
+        vars_.erase(iter++);
+      }
+    }
+  }
+}
```

Comment:
I prefer that not add `ClearWithPreserve` here, because scope provides `EraseVars` and `DropKids`, and `ClearWithPreserve` can be replaced by `EraseVars` and `DropKids`. **We should ensure that the function of the interface of scope is orthogonal.**

Reply:
I agree. But `Scope` does not have a method like `EraseVarsExcept`. How about add this method?